src/main/java/com/aws/greengrass/mqtt/bridge/clients/IoTCoreClient.java

```diff
@@ -143,89 +121,260 @@ public void publish(com.aws.greengrass.mqtt.bridge.model.MqttMessage message) {
         return (com.aws.greengrass.mqtt.bridge.model.MqttMessage) message.toMqtt();
     }
+
+    // for unit testing
+    Set<String> getSubscribed() {
+        return subscriptionManager.getSubscribed();
+    }
+
+    // for unit testing
+    Set<String> getToSubscribe() {
+        return subscriptionManager.getToSubscribe();
+    }
+
+    // TODO factor out so can be used by local v5 client
+    class SubscriptionManager {
+        private final Object subscriptionLock = new Object();
+        private final Set<String> subscribed = new HashSet<>();
+        private final Set<String> toSubscribe = new HashSet<>();
+        private final AtomicBoolean online = new AtomicBoolean(true);
+        private final AtomicBoolean shutdown = new AtomicBoolean(false);
+
+        /**
+         * Stop accepting new subscription requests, cancel all in-progress subscriptions,
+         * and unsubscribe.
+         */
+        public void shutdown(long timeout, TimeUnit unit) {
+            if (!shutdown.compareAndSet(false, true)) {
+                return;
+            }
+            try {
+                subscriptionManager.unsubscribeAll().get(timeout, unit);
+            } catch (ExecutionException e) {
+                LOGGER.atWarn().cause(Utils.getUltimateCause(e)).log("unable to unsubscribe from topics");
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            } catch (TimeoutException e) {
+                LOGGER.atWarn().cause(Utils.getUltimateCause(e)).log("timed-out when unsubscribing from topics");
+            }
+        }
+
+        /**
+         * Subscribe to all known topics. All previously failed/not-made subscriptions will
+         * attempt to subscribed again. If a topic has already been subscribed to, it will not be
+         * resubscribed.
+         */
+        public CompletableFuture<Void> subscribeAll() {
+            Set<String> topicsToSubscribe;
+            synchronized (subscriptionLock) {
+                topicsToSubscribe = new HashSet<>(toSubscribe);
+                topicsToSubscribe.removeAll(subscribed);
+            }
+            return subscribe(topicsToSubscribe);
+        }
+
+        public CompletableFuture<Void> subscribe(Collection<String> topics) {
+            CompletableFuture<Void> result = CompletableFuture.supplyAsync(() -> null, executorService);
+            for (String topic : topics) {
+                result.thenComposeAsync(unused -> {
+                    if (canSubscribe()) {
+                        return subscribeWithRetry(topic);
+                    } else {
+                        return CompletableFuture.completedFuture(null);
+                    }
+                }, executorService);
+            }
+            return result;
+        }
+
+        private CompletableFuture<Void> subscribeWithRetry(String topic) {
+            return IoTCoreClient.this.subscribe(topic).thenComposeAsync(resp -> {
```

Comment:
I don't see a reason why you need to compose async. Just composing should be fine. It will run on the MQTT client's thread, so be aware of this, but it should be fine?